src/v/raft/persisted_stm.cc

```diff
@@ -545,17 +545,26 @@ ss::future<> persisted_stm_base<BaseT, T>::start() {
 
     if (maybe_snapshot) {
         stm_snapshot& snapshot = *maybe_snapshot;
-
         auto next_offset = model::next_offset(snapshot.header.offset);
         if (next_offset >= _raft->start_offset()) {
-            vlog(
-              _log.debug,
-              "start with applied snapshot, set_next {}",
-              next_offset);
-            co_await apply_local_snapshot(
+            auto snapshot_applied = co_await apply_local_snapshot(
               snapshot.header, std::move(snapshot.data));
+            if (snapshot_applied == local_snapshot_applied::yes) {
+                vlog(
+                  _log.debug,
+                  "start with applied snapshot, set_next {}",
+                  next_offset);
+                _last_snapshot_offset = snapshot.header.offset;
+            } else {
+                next_offset = _raft->start_offset();
```

Comment:
I wonder if we should leave the next offset as is here. If we set it to the start offset, we will skip application of the raft snapshot, which is not what we want.

Reply:
good catch, done.